src/platform/ESP32/MdnsBasicPacketFilters.h

```diff
@@ -0,0 +1,112 @@
+ *        are accumulated, and drops those that would cause crossing of the threshold.
+ *
+ */
+class DropIfTooManyQueuedMdnsPacketFilter : public chip::Inet::EndpointQueueFilter
+{
+    // Repeated here since there are no accessible headers with this
+    static constexpr uint16_t kMdnsPort = 5353u;
+
+public:
+    /**
+     * @brief Initialize the packet filter with a starting limit
+     *
+     * @param maxQueuedPackets - max number of pending-in-queue mDNS packets not yet processed
+     */
+    DropIfTooManyQueuedMdnsPacketFilter(size_t maxQueuedPackets) : mMaxQueuedPackets(maxQueuedPackets) {}
+
+    FilterOutcome FilterBeforeEnqueue(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                      const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        // WARNING: This is likely called in a different context than `FilterAfterDequeue`. We use an atomic for the counter.
+
+        // Non-mDNS is not filtered
+        VerifyOrReturnValue(pktInfo.DestPort == kMdnsPort, FilterOutcome::kAllowPacket);
+
+        if (pktInfo.DestAddress.IsIPv4() && pktInfo.DestPort == kMdnsPort)
+        {
+            if (!mStopLogging)
+            {
+                ChipLogError(Discovery, "Started dropping mDNS from storm.");
+                mStopLogging = true;
+            }
+            return FilterOutcome::kDropPacket;
+        }
+
+        if (mNumQueuedMdnsPackets.load() >= mMaxQueuedPackets)
+        {
+            if (!mStopLogging)
+            {
+                ChipLogError(Discovery, "Started dropping mDNS from storm.");
+                mStopLogging = true;
+            }
+            return FilterOutcome::kDropPacket;
+        }
+
+        ++mNumQueuedMdnsPackets;
+
+        return FilterOutcome::kAllowPacket;
+    }
+
+    FilterOutcome FilterAfterDequeue(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                     const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        // WARNING: This is likely called in a different context than `FilterBeforeEnqueue`. We use an atomic for the counter.
+
+        // Non-mDNS is ignored
+        VerifyOrReturnValue(pktInfo.DestPort == kMdnsPort, FilterOutcome::kAllowPacket);
+
+        --mNumQueuedMdnsPackets;
+        if (mNumQueuedMdnsPackets.load() == 0)
+        {
+            if (mStopLogging)
+            {
+                mStopLogging = false;
+                ChipLogError(Discovery, "Resolved mDNS storm.");
+            }
+        }
+
+        // We always allow the packet and just do accounting, since all dropping is prior to queue entry.
+        return FilterOutcome::kAllowPacket;
+    }
+
+    void SetMaxQueuedPacketsLimit(size_t maxQueuedPackets) { mMaxQueuedPackets = maxQueuedPackets; }
+
+protected:
+    std::atomic_size_t mNumQueuedMdnsPackets = { 0 };
+    size_t mMaxQueuedPackets                 = 0;
```

Comment:
```suggestion
    const size_t mMaxQueuedPackets                 = 0;
```
so it's clear why a non-atomic thing is ok here.

Reply:
added this change